Feature/aerospike cache improvements

roimenashe opened the pull request.

Aloren commented:
thank you, I will take a look today. unfortunately i was on a sick leave today 🙁

Aloren commented on src/main/java/org/springframework/data/aerospike/cache/AerospikeCache.java:
```diff
@@ -1,114 +1,144 @@
+						  AerospikeConverter aerospikeConverter,
+						  String namespace,
+						  String set,
+						  int expiration) {
```
tip: `expiration` better to rename to `expirationInSeconds`

Aloren commented on src/main/java/org/springframework/data/aerospike/cache/AerospikeCache.java:
```diff
@@ -1,114 +1,144 @@
 	}
 
-	protected Key getKey(Object key){
+	private Key getKey(Object key){
```
one more enhancement for the future that we can give to the end user: add ability to provide custom converter for the user-key part (`key.toString()` -> `converter.findConverter(key, Value.class)`)

roimenashe replied:
I think it’s less relevant for this PR, but as you said it’s a nice to have and maybe we should add it as "enhancement" issue.

Aloren commented on src/main/java/org/springframework/data/aerospike/cache/AerospikeCache.java:
```diff
@@ -1,114 +1,144 @@
+					put(key, value);
+				}
+			} catch (Exception e) {
+				log.warn("valueLoader exception", e);
```
according to the documentation from the interface:
```
If the valueLoader throws an exception, it is wrapped in a Cache.ValueRetrievalException
```
we should not swallow an exception and instead should wrap it into `Cache.ValueRetrievalException` and rethrow

Aloren replied:
also let's replace `catch (Exception e) {` with `catch (Throwable e) {`, as catching Exception type does not catch all exceptions

Aloren commented on src/main/java/org/springframework/data/aerospike/cache/AerospikeCache.java:
```diff
@@ -1,114 +1,144 @@
 	public ValueWrapper putIfAbsent(Object key, Object value) {
-		Record record = client.operate(this.createOnly, getKey(key), Operation.put(new Bin(VALUE, value)), Operation.get(VALUE));
-		return toWrapper(record);
+		serializeAndPut(createOnly, key, value);
```
if value already exists -- shouldn't we just return it? `RecordExistsAction.CREATE_ONLY` will fail with exception if data is present

Aloren commented on src/main/java/org/springframework/data/aerospike/cache/AerospikeCacheManager.java:
```diff
@@ -124,28 +190,16 @@ protected Cache getMissingCache(String cacheName) {
-	protected AerospikeCache createCache(String cacheName) {
-		return new AerospikeSerializingCache(cacheName);
+	private AerospikeCache createCache(String cacheName) {
+		return new AerospikeCache(aerospikeClient, aerospikeConverter, cacheName, setName, defaultTimeToLive);
```
```
AerospikeCache(IAerospikeClient client,
						  AerospikeConverter aerospikeConverter,
						  String namespace,
						  String set,
						  int expiration)
```
is `cacheName` always a `namespace`?

roimenashe replied:
It seems so,
The only confusion is in getName() in AerospikeCache and the lookup in AerospikeCacheManager, both are using a combination of namespace and set name, this is the original code - I just rearranged it.

roimenashe replied:
In the original code - yes, but, we changed the design a bit to be more simple and to solve some issues we had.

Now, a cache has a name that is mapped to a cache configuration - the cache configuration has a set and a namespace (along with some other config values).

The default namespace is "test" and the default set is "aerospike" - in tests the default namespace is "TEST" to match the Playtika embedded namespace value.

Aloren commented on src/main/java/org/springframework/data/aerospike/cache/AerospikeCacheManager.java:
```diff
@@ -124,28 +190,16 @@ protected Cache getMissingCache(String cacheName) {
-			}
-			return null;
-		}
+	protected Cache getAerospikeCache(String cacheName) {
```
is it possible to remove getAerospike named methods from this class? it already has getCache, lookupCache and one more set of almost the same methods makes it difficult to use

roimenashe replied:
In the new design there are no custom "Aerospike" methods - such as getAerospikeCache.

Aloren commented on src/main/java/org/springframework/data/aerospike/cache/AerospikeCacheManager.java:
```diff
@@ -41,78 +34,151 @@
+	public AerospikeCacheManager(IAerospikeClient aerospikeClient,
+								 AerospikeConverter aerospikeConverter,
+								 Collection<String> cacheNames,
+								 int defaultTimeToLive) {
```
defaultTimeToLiveInSeconds
or we can use Duration type

Aloren commented:
@reugn @roimenashe i've left some comments. i think, we can make fixes to the current implementation in this PR, what do you think?

reugn commented:
@Aloren agree.

roimenashe commented:
@Aloren There was a design change to overcome some issues we had and for better readability, please review the latest changes.

Regarding your comments, i've addressed them in the new design and replied.

Thanks

Aloren commented on src/main/java/org/springframework/data/aerospike/cache/AerospikeCacheManager.java:
```diff
@@ -41,111 +36,107 @@
+	public AerospikeCacheManager(IAerospikeClient aerospikeClient,
+								 AerospikeConverter aerospikeConverter,
+								 AerospikeCacheConfiguration defaultCacheConfiguration,
+								 String... initialCacheNames) {
```
varargs is nice to use in test scenarios, unfortunately it is not very convenient for production scenarios. for example, when such things are provided via configurations. i would suggest using Collection instead or List in particular

reugn replied:
I don't think this is critical.

Aloren replied:
Of course this is not critical. From this point of view any changes for usability of the end user are not critical ;)

roimenashe replied:
Changed varargs to a List.

Aloren commented on src/main/java/org/springframework/data/aerospike/cache/AerospikeCacheManager.java:
```diff
@@ -41,111 +36,107 @@
+	public AerospikeCacheManager(IAerospikeClient aerospikeClient,
+								 AerospikeConverter aerospikeConverter,
+								 AerospikeCacheConfiguration defaultCacheConfiguration,
+								 Map<String, AerospikeCacheConfiguration> initialCacheConfiguration) {
```
I'm not sure that it is obvious from the documentation what stands behind the key of the `initialCacheConfiguration`. Is it a cache name?

roimenashe replied:
Yes the key is a cache name and the value is an AerospikeCacheConfiguration, I added a bit more description I hope its clear now.

Aloren commented on src/main/java/org/springframework/data/aerospike/cache/AerospikeCacheManager.java:
```diff
@@ -41,111 +36,107 @@
-		this.setName = setName;
-		this.configuredCacheNames = new LinkedHashSet<String>(cacheNames);
+		this.defaultCacheConfiguration = (defaultCacheConfiguration != null ? defaultCacheConfiguration : new AerospikeCacheConfiguration());
+		this.initialCacheConfiguration = (initialCacheConfiguration != null ? initialCacheConfiguration : new LinkedHashMap<>());
```
probably it is `perCacheConfiguration` or smth like that?

roimenashe replied:
I changed the name to initialPerCacheConfiguration.
Basically it's a map of pre-configured cache names and a matching configuration that will be created at the start of the application (it is not an image of which caches we currently have - if a cache is created at runtime it will not be added to this map).

Aloren commented on src/main/java/org/springframework/data/aerospike/cache/AerospikeCacheManager.java:
```diff
@@ -41,111 +36,107 @@
-	protected Cache lookupAerospikeCache(String name) {
-		return lookupCache(name + ":" + setName);
+	private AerospikeCache createCache(String name, @Nullable AerospikeCacheConfiguration cacheConfiguration) {
+		return new AerospikeCache(name, aerospikeClient, aerospikeConverter, cacheConfiguration != null ? cacheConfiguration : defaultCacheConfiguration);
```
what is the purpose of giving end users an api for providing a map with null values? Having null values in collections is considered nowadays a design flaw, so I would suggest prohibiting passing null values

Aloren commented on src/test/java/org/springframework/data/aerospike/cache/AerospikeCacheManagerIntegrationTests.java:
```diff
@@ -52,6 +125,35 @@ public void shouldEvictCache() {
+
+        assertThat(cachingComponent.getNoOfCalls()).isEqualTo(1);
+
+        Thread.sleep(3000);
```
for such cases you can also use a nice library https://github.com/awaitility/awaitility

reugn replied:
This is a good one. @roimenashe there is a lot of `Thread.sleep` in the tests. Let's open an issue to replace them with `awaitility` or similar. But this should not be a blocker for this PR.

Aloren replied:
this is just a tip

roimenashe replied:
I agree, after looking at awaitility it looks like a better solution - I think we should separate it to another PR that replaces Thread.sleep usages with awaitility in general (not only in the cache package).

reugn commented on src/main/java/org/springframework/data/aerospike/cache/AerospikeCacheManager.java:
```diff
@@ -41,111 +36,107 @@
  * @author Venil Noronha
  */
-//TODO: extract this to the separate repository aerospike-spring-cache
+@Builder
```
It doesn't expose the fields of the base class.

Aloren replied:
As you wish of course, but according to this constructor:
`public AerospikeCacheManager(IAerospikeClient aerospikeClient,
								 AerospikeConverter aerospikeConverter,
								 AerospikeCacheConfiguration defaultCacheConfiguration,
								 Map<String, AerospikeCacheConfiguration> initialCacheConfiguration)`
how the internals of `AerospikeCacheManager` are going to be setup with `builder`? Exposing too much internals to the end user will give you less possibilities to change them in the future, so what I'm warning you is from tying your hands.

roimenashe replied:
I removed the builder pattern and added the missing relevant constructors - I though its also made the code a bit unclear having a builder pattern with 4-5 constructors.

roimenashe commented on src/main/java/org/springframework/data/aerospike/cache/AerospikeCacheConfiguration.java:
```diff
@@ -0,0 +1,28 @@
+
+    protected static final String DEFAULT_NAMESPACE = "test";
+    protected static final String DEFAULT_SET_NAME = "aerospike";
+    protected static final int DEFAULT_EXPIRATION = -1;
```
There are no more "explicit" default values after the latest refactor.
A user must specify namespace when configuring a cache manager.
"default" set is null (writing directly to an Aerospike namespace).
"default" expiration is 0 (use Aerospike server's default - as described in the WritePolicy of Aerospike java client.

roimenashe commented:
@Aloren I addressed your latest comments and added a little refactor in the configuration area, also removed a few constructors from the cache manager since we felt they are adding more confusing than helpful.

Please review the latest changes and approve\reply, Thank you.

Aloren commented on src/main/java/org/springframework/data/aerospike/cache/AerospikeCacheConfiguration.java:
```diff
@@ -0,0 +1,26 @@
+public class AerospikeCacheConfiguration {
+    private final String namespace;
+    private String set; // null: Default set is null meaning write directly to the namespace.
+    private int expirationInSeconds; // 0: Default to namespace configuration variable "default-ttl" on the server.
```
it doesn't look like it is possible to change `AerospikeCacheConfiguration` after initialization, so we can make all fields final
